### PR TITLE
check meta, treewide

### PR DESCRIPTION
**Pull request by vcunat**
#### Plan:

- [x] wait a bit for feedback (some parts might be disputable)
- [x] backport to 17.09
- [ ] [integrate into borg checks](https://github.com/grahamc/ofborg/pull/16)

**?**: add into the tarball job, simple pre-push hook for direct pushers, more meta checks, ...

### CODE REVIEW

**Comment by vcunat** [comment]
and /cc @copumpkin who started `checkMeta`.

**Comment by vcunat** [comment]
For now I used ``nix-env -f checkMeta.nix -qa --no-name --out-path --show-trace >/dev/null`` with [some code adapted from rebuild-amount.sh](http://sprunge.us/PHAP).

**Comment by 7c6f434c** [comment]
For some weird GitHub bug I cannot leave a line comment; wouldn't it be more consistent to also add the current platform to the message for Parallels tools?

I guess perfect consistency w.r.t. system/platform and colon/no-colon is too annoying to ensure across all the changes (for little gain).

In general looks good, but not sure if GitHub hides some part of diff because it is too large…

File: pkgs/applications/taxes/aangifte-2006/default.nix [hunk 1]
```diff
@@ -5,15 +5,15 @@ stdenv.mkDerivation {
   builder = ./builder.sh;
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2006_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2006_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2006/default.nix]: this one isn't right

File: pkgs/applications/taxes/aangifte-2007/default.nix [hunk 2]
```diff
@@ -5,7 +5,7 @@ stdenv.mkDerivation {
   builder = ./builder.sh;
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2007_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2007_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2007/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2008/default.nix [hunk 3]
```diff
@@ -6,7 +6,7 @@ stdenv.mkDerivation {
   builder = ./builder.sh;
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2008_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2008_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2008/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2009/default.nix [hunk 4]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2009-1";
   
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2009_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2009_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2009/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2010/default.nix [hunk 5]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2010-1";
   
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2010_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2010_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2010/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2011/default.nix [hunk 6]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2011-1";
   
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2011_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2011_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2011/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2012/default.nix [hunk 7]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2012-1";
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2012_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2012_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2012/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2013-wa/default.nix [hunk 8]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2013-wa";
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/wa2013_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/wa2013_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2013-wa/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2013/default.nix [hunk 9]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2013-1";
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2013_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2013_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2013/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2014-wa/default.nix [hunk 10]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2014-wa";
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/wa2014_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/wa2014_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2014-wa/default.nix]: url / homepage confusion

File: pkgs/applications/taxes/aangifte-2014/default.nix [hunk 11]
```diff
@@ -4,7 +4,7 @@ stdenv.mkDerivation {
   name = "aangifte2014-1";
 
   src = fetchurl {
-    url = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2014_linux.tar.gz;
+    homepage = http://download.belastingdienst.nl/belastingdienst/apps/linux/ib2014_linux.tar.gz;
```
Comment (grahamc) [inline pkgs/applications/taxes/aangifte-2014/default.nix]: url / homepage confusion

**Review by grahamc — COMMENTED** [review]
looks great other than these very minor issues :)

File: pkgs/applications/taxes/aangifte-2006/default.nix (hunk 1 above)
Comment (vcunat) [reply pkgs/applications/taxes/aangifte-2006/default.nix]: Oh, aangifte-* were the only not done fully manually and they got wrong without me noticing :-/

**Review by grahamc — APPROVED** [review]
I think I don't fully grok the implications around abort vs. throw but I strongly approve of this!

**Comment by vcunat** [comment]
We could add something like `stdenv.throwSystem` or frameworkize this in some other way :-)

File: pkgs/stdenv/generic/check-meta.nix [hunk 12]
```diff
@@ -126,7 +126,7 @@ let
       '';
 
   throwEvalHelp = { reason , errormsg ? "" }:
-    throw (''
+    (if reason != "unknown-meta" then throw else (x : builtins.trace x true)) (''
```
Comment (grahamc) [inline pkgs/stdenv/generic/check-meta.nix]: I'd be tempted to make this configurable so we can highlight or ignore certain conditions depending on what we're testing for:

```diff
diff --git a/pkgs/stdenv/generic/check-meta.nix b/pkgs/stdenv/generic/check-meta.nix
index 03a85633ed6..308a0f370c7 100644
--- a/pkgs/stdenv/generic/check-meta.nix
+++ b/pkgs/stdenv/generic/check-meta.nix
@@ -125,11 +125,12 @@ let
 
       '';
 
-  throwEvalHelp = { reason , errormsg ? "" }:
+  handleEvalIssueDefault = { reason , errormsg ? "" }:
     throw (''
       Package ‘${attrs.name or "«name-missing»"}’ in ${pos_str} ${errormsg}, refusing to evaluate.
 
       '' + ((builtins.getAttr reason remediation) attrs));
+  handleEvalIssue = x: (config.handleEvalIssue or handleEvalIssueDefault) x;
 
   metaTypes = with lib.types; rec {
     # These keys are documented
@@ -192,7 +193,7 @@ let
   validityCondition =
          let v = checkValidity attrs;
          in if !v.valid
-           then throwEvalHelp (removeAttrs v ["valid"])
+           then handleEvalIssue (removeAttrs v ["valid"])
            else true;
 
 in
```

Comment (grahamc) [reply pkgs/stdenv/generic/check-meta.nix]: If we play our cards right, we could take `evaluation aborted with the following error message: ‘Package ‘aangifte2006-1’ in /home/grahamc/projects/nixpkgs/pkgs/applications/taxes/aangifte-2006/default.nix:15 has an invalid meta attrset:
` parse the path an line, and leave in-line feedback on PRs..........

Comment (vcunat) [reply pkgs/stdenv/generic/check-meta.nix]: I'm also tempted by this.

**Comment by oxij** [comment]
If I were you I would squash 374eba46ee9488e242b0dae5cec163bcc8cf27bf
and ba940713764d43bb3f735d4c74fbc5162e83726a (they LGTM) and cherry-pick
the result to master and 17.09 as soon as possible. The leftover diff
would get much smaller and users which don't really care about the
frameworkization part (e.g. me) would instantly get happy.

File: pkgs/stdenv/generic/check-meta.nix (hunk 12 above)
Comment (vcunat) [reply pkgs/stdenv/generic/check-meta.nix]: Oh, I think this will be a good way to clean https://hydra.nixos.org/jobset/nixpkgs/trunk#tabs-errors 👍

**Pull request merged**